### ISSUE TITLE
Question / Comment: Possible to extract sound annotations from PDF?

### ISSUE THREAD

**Issue by Saya47** [pymupdf/PyMuPDF#683]
Hello is it possible to extract sound annotations from a PDF to MP3 files?

If so, can anyone help me with the script?

Thanks very much!

**Comment by JorjMcKie** [comment]
Sound annotations are currently not supported. It shouldn't be too difficult to implement a method which extracts the binary stream that the sound annotation references via the PDF key ``/Sound``. Once this is implemented, that stream would land in a ``bytes`` Python object, which subsequently could be saved as a file.
What type of sound this is (MP§ or whatever), would however be beyond scope and left to some downstream logic.

**Comment by JorjMcKie** [comment]
Well, I was a bit too pessimistic:
There are a number of optional parameters that may accompany the sheer binary sound stream. If / when this feature is being implemented, the return of the sound extraction method would probably be a Python dictionary, containing some or all of the fields explained on page 783 of https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/pdf_reference_1-7.pdf.

**Comment by JorjMcKie** [comment]
Can you please send me a (small!) PDF example with a sound annot?

**Comment by Saya47** [comment]
Of course! Here's one I created: first page has two sound annotations and second page has one sound annotation:

[Sound annot example.pdf](https://github.com/pymupdf/PyMuPDF/files/5359430/Sound.annot.example.pdf)


Thanks very much for looking into this.

**Comment by JorjMcKie** [comment]
I have a first version ready. If you tell me your config (OS, bitness, Python), I will point you to a location for downloading a pre-version wheel.
The API looks like this:
```python
>>> annot = page.firstAnnot
>>> sound = annot.soundGet()  # raises exception if not a sound
>>> for k, v in sound.items():
	print(k, "=", v if k != "stream" else len(v))
rate = 48000.0
channels = 2
bps = 16
encoding = Signed
stream = 677664
>>> 
```
``sound["stream"]`` contains the sound file content (``bytes`` object), the other keys are sampling rate, number of channels, bits per sample, encoding.
I haven't implemented the case where the sound is represented as a separate file in the PDF. None of your 3 example sounds is like that. So if you need that support too, I would need an example for this as well. So far an exception is raised if this is the case.

**Comment by Saya47** [comment]
That's perfect! Thank you! I'm on Windows 10 X64 1803 and running Python 3.8.6 X64.

I personally don't need the ability to extract audios if they are attachments because they're extractable with Adobe Acrobat, but in case you would like to add support for that just in case, here's an example with 3 audio attachments:

[PDF with 3 audio attachments.pdf](https://github.com/pymupdf/PyMuPDF/files/5360941/PDF.with.3.audio.attachments.pdf)

**Comment by JorjMcKie** [comment]
That's a PDF with just file attachment annotations - which already is supported: you can extract them no problem.
No what I was eluding to are **_sound_** annotations which point to an embedded file.
But anyway no problem. I will postpone this until someone wants that.
Next post will be the wheel for you ...

**Comment by JorjMcKie** [comment]
[PyMuPDF-1.18.1-cp38-cp38-win_amd64.zip](https://github.com/pymupdf/PyMuPDF/files/5360970/PyMuPDF-1.18.1-cp38-cp38-win_amd64.zip)
Install this like so:
1. rename extension back to "whl" (from "zip") - GitHub won't allow otherwise
2. `python -m pip install -U PyMuPDF-1.18.1-cp38-cp38-win_amd64.whl`

**Comment by JorjMcKie** [comment]
hope it works - I am currently doing other things in parallel, so errors easily happen ... but good luck!

**Comment by Saya47** [comment]
Hello! It doesn't seem to be working as expected, the result is harsh static noise!

I wrote a little script here which converts the raw files to mp3. And I tested it on our Annotations PDF example. 

But even when I import the raw files to an audio editing software like Adobe Audition, and choose sampling rate 48k and 2 channels and PCM 16 bit Signed, it's the same result! 

```
import fitz, sys, os, subprocess
assert len(sys.argv) == 2, "need filename as parameter"
ifile = sys.argv[1]
doc = fitz.open(sys.argv[1])
for page in doc:
    annot = page.firstAnnot
    sound = annot.soundGet()  # raises exception if not a sound
    for k, v in sound.items():
        print(k, "=", v if k != "stream" else len(v))
    ofile = ifile + str(page.number) + ".raw"
    fout = open(ofile,"wb") # Write binary
    fout.write(sound["stream"])
    fout.close()
    ofileffmpeg = ofile + ".mp3"
    print(ofileffmpeg)
    if "channels" in sound:
        subprocess.call(['ffmpeg', '-hide_banner', '-f', 'u16le', '-ar', str(sound["rate"]), '-ac', str(sound["channels"]), '-i', str(ofile), str(ofileffmpeg)], shell=True)
    else:
        subprocess.call(['ffmpeg', '-hide_banner', '-f', 'u16le', '-ar', str(sound["rate"]), '-ac', '1', '-i', str(ofile), str(ofileffmpeg)], shell=True)
```

**Comment by Saya47** [comment]
Please wait, I need to do more tests, there might be an issue with the PCM being 32 bit I'm not sure.

**Comment by JorjMcKie** [comment]
There is an iterator for the page:
```python
for page in doc:
    for annot in page.annots(types=[fitz.PDF_ANNOT_SOUND]):  # optional subselection for annot types- list or tuple
        # do something with the annot
```
When I extract the annot sound stream, I use a function which decomresses the binary stream. That could be changed to return that stream undecompressed ... you could then decompress as desired or not.

**Comment by Saya47** [comment]
Oh I see, that could be very useful, I'd really appreciate it if you could add the option.

**Comment by JorjMcKie** [comment]
ok, give me a few minutes.
``sound["image"]`` will then extract the sound binary "raw" .

**Comment by Saya47** [comment]
Wait that's not the problem! I figured it out! It's the byte order, they need to be imported as Big Endian and they will play fine!

**Comment by Saya47** [comment]
![Capture](https://user-images.githubusercontent.com/35542834/95683433-ffb7df00-0bf7-11eb-9673-59fbac552357.PNG)

Sorry for the trouble.
And thanks a lot <3

**Comment by JorjMcKie** [comment]
no problem!
so you are fine as it is?

**Comment by Saya47** [comment]
Yes sir! Thank you so very much!

**Comment by JorjMcKie** [comment]
New version 1.18.1 uploaded to PyPI.

**Comment by CharlieMcVicker** [comment]
I know I'm late to the party on this, but I had a PDF with embedded audio tied to radio button widgets and the above did not work for me. Instead, I went with the following approach, which may need to be changed if your audio is in a different format.

```python

import fitz
from scipy.io.wavfile import write as write_wav
import numpy as np 

doc = fitz.open("Tsugatsalesdi Tsunvtsidv.pdf")

fileNo = 0
subtypes = set()
xreflen = doc.xref_length()
dt = np.dtype('int16')
dt = dt.newbyteorder('>')
for xref in range(1, xreflen):
    if doc.xref_is_stream(xref):
        if 'Type' in doc.xref_get_keys(xref):
            name, type = doc.xref_get_key(xref, 'Type')
            if type == '/Sound':
                stream = doc.xref_stream(xref)
                filename = f"dump_file_{fileNo}.wav"
                channels = int(doc.xref_get_key(xref, 'C')[1])
                bitdepth = int(doc.xref_get_key(xref, 'B')[1])
                samplerate = int(doc.xref_get_key(xref, 'R')[1])
                signed = doc.xref_get_key(xref, 'E')[1] == '/Signed'
                if not signed or bitdepth != 16:
                    raise Exception("Unsupported format")
                arr = np.frombuffer(stream, dtype=dt).reshape((-1, channels))
                write_wav(f"dump_file_{fileNo}.wav", samplerate, arr)
                fileNo += 1
```
